### PR TITLE
Ability to turn off publish retries if broker is disconnected

### PR DESCRIPTION
**Pull request by raybooysen**
This is a quick WIP commit to discuss options on how to achieve this.  I've added an option to the configuration so that a consumer can turn off the retries. 

Some thoughts:

1. What should the behaviour be if it's disconnected, retries are turned off and a message is published (throw? drop?)
2. What should we do about the unit tests?  I'm loathe to copy and paste the set of tests verifying that tests that queuing is working, but seems the easiest.  Thoughts?

### CODE REVIEW

**Comment by kum-deepak** [comment]
Many thanks for this work. I have following inputs:

- It would be better to add the option in the `publish` call. That way one can choose some messages for retires and some without (use cases in our company would likely follow this pattern).
- Copying and modifying existing test cases should be a good approach. Later, only if needed, these can be refactored.

Answer to your questions from #6:

- Usually in the world of STOMP sending should only fail if connection is not there. Other cases the broker is supposed report errors later using an ERROR frame. In addition, a compliant broker will terminate the connection after ERROR frame (in practice some brokers do not).
- It is a good idea to defensively handle attempt to send each of the messages.
- When the retry is off, both silently dropping or raising exception are perfectly valid. In general there may be other reasons the broker may drop the message (e.g., because of a failure), so, as a good practice clients should be ready to handle that. However, many users are likely to be confused. So, I propose raising an exception, say `NotConnectedException`.
- Your understanding about the current functionality is correct. `retry` is probably not entirely correct. Feel free to suggest an alternate.

**Comment by raybooysen** [comment]
Pushed a few changes, what do you think?

**Comment by kum-deepak** [comment]
This looks quite good already. My feedback:

- I realized hard way that option flags are better kept as Hash arguments (or an Interface - for type safety). This allows much neater code structure in future if more options need to be added.
- I am getting aligned towards an alternate approach - the first parameter of this call is already a Hash (https://github.com/stomp-js/stompjs/blob/master/src/types.ts#L48). This can be extended (say `IExtendedPublishParams`) to add this extra option. What do you think?
- Will you consider renaming the flag to `retryIfDisconnected` or `enqueIfDisconnected`?

I am also working on updating the contribution guide (#95). Your feedback is welcome: https://github.com/stomp-js/rx-stomp/pull/95/files?file-filters%5B%5D=.md&owned-by%5B%5D=

File: src/rx-stomp.ts [hunk 1]
```diff
@@ -370,16 +370,21 @@ export class RxStomp {
...
-  public publish(parameters: publishParams): void {
+  public publish(parameters: publishParams, retryOnDisconnect?: boolean): void {
     if (this.connected()) {
       this._stompClient.publish(parameters);
-    } else {
+    } else if (retryOnDisconnect != null && retryOnDisconnect) {
       this._debug(`Not connected, queueing`);
       this._queuedMessages.push(parameters);
+    } else {
+      throw new Error('Cannot publish while connection is not connected');
```
Comment (kum-deepak) [inline src/rx-stomp.ts]:
```suggestion
      throw new Error('Cannot publish while broker is not connected');
```

Just reads better.

Comment (raybooysen) [reply src/rx-stomp.ts]: Done

**Comment by raybooysen** [comment]
> This looks quite good already. My feedback:
> 
> * I realized hard way that option flags are better kept as Hash arguments (or an Interface - for type safety). This allows much neater code structure in future if more options need to be added.
> * I am getting aligned towards an alternate approach - the first parameter of this call is already a Hash (https://github.com/stomp-js/stompjs/blob/master/src/types.ts#L48). This can be extended (say `IExtendedPublishParams`) to add this extra option. What do you think?

Done, but I'd unsure on naming of files in the code-base.  Can you check?

> * Will you consider renaming the flag to `retryIfDisconnected` or `enqueIfDisconnected`?
> 
Done

> I am also working on updating the contribution guide (#95). Your feedback is welcome: https://github.com/stomp-js/rx-stomp/pull/95/files?file-filters%5B%5D=.md&owned-by%5B%5D=

**Comment by kum-deepak** [comment]
I have made renames and documentation changes. Please check https://github.com/stomp-js/rx-stomp/commits/develop and see if all your changes are intact.

**Comment by kum-deepak** [comment]
With your confirmation, I will close this PR.

**Comment by raybooysen** [comment]
Fine by me.

**Pull request closed without merging**